Add support for single-threaded CREATE INDEX.

wuwenw opened the pull request:
Similart to #990 but now it's on tpl2. Currently only work in single thread.

There are three parts: 
1. create index through ddl_executor
2. sequantial scan on the table
3. insert tuples into new index

wuwenw commented:
For now, there are three testcases written in cpp
1. CREATE INDEX test_index on index_test_table(colA), Select colA from table and index to verify
2. CREATE INDEX test_index on index_test_table(colE), Select * from table and index to verify
3. CREATE INDEX test_index on index_test_table(colA, colB), Select colA, colB, colC, colD from table and index to verify

mbutrovich commented:
For context to reviewers, @amlatyrngom took 2 passes through #990 so I trust the general design at this point. I think we're mostly going to be looking at best practices, style, etc.

lmwnshn commented on src/execution/compiler/operator/index_create_translator.cpp:
```diff
@@ -0,0 +1,228 @@
+}
+
+void IndexCreateTranslator::DeclareTablePR(FunctionBuilder *function) const {
+  // var table_pr = @InitTablePR()
```
Typo: @initTablePR

lmwnshn commented on src/include/execution/sql/ddl_executors.h:
```diff
@@ -67,6 +67,14 @@ class DDLExecutors {
                                   common::ManagedPointer<catalog::CatalogAccessor> accessor);
 
+  /**
+   * @param node node to executed
```
Typo: node to execute

lmwnshn commented on src/execution/compiler/operator/index_create_translator.cpp:
```diff
@@ -0,0 +1,228 @@
+void IndexCreateTranslator::CreateIndex(FunctionBuilder *function) const {
+  auto plan_node = common::ManagedPointer<const planner::CreateIndexPlanNode>(
+      dynamic_cast<const planner::CreateIndexPlanNode *>(Op()));
+  auto result = sql::DDLExecutors::CreateIndexExecutor(
```
Do we want to create the index while generating TPL code? It seems more correct to have a TPL builtin, @createIndex(...) which returns a boolean. You would then use the boolean here.

wuwenw replied:
Now index is created before generating TPL code

lmwnshn commented on src/execution/sema/sema_builtin.cpp:
```diff
@@ -3059,6 +3080,8 @@ void Sema::CheckBuiltinCall(ast::CallExpr *call) {
       break;
     }
     case ast::Builtin::StorageInterfaceInit:
+    case ast::Builtin::InitTablePR:
```
Is there a reason you're using table PRs instead of just getting it directly from the VPI?

`@prSetIntNull(update_pr, 0, @vpiGetIntNull(vpi, 0))`
vs
`@prSetIntNull(index_pr, 0, @prGetIntNull(table_pr, 0))`

If we go with the table_pr approach, this requires an extra allocation and deallocation for the table_pr with every storage interface instantiation.

wuwenw replied:
Now it follows @prSetIntNull(update_pr, 0, @vpiGetIntNull(vpi, 0))

lmwnshn commented on src/execution/sql/ddl_executors.cpp:
```diff
@@ -106,6 +106,12 @@ bool DDLExecutors::CreateIndexExecutor(const common::ManagedPointer<planner::Cre
                      *(node->GetSchema()));
 }
 
+bool DDLExecutors::CreateIndexExecutor(const common::ManagedPointer<const planner::CreateIndexPlanNode> node,
```
This looks identical to the above code. Can you un-const it where it is being called? If not, can you remove the other version?

wuwenw replied:
removed

lmwnshn commented on test/optimizer/index_nested_loops_join_test.cpp:
```diff
@@ -46,7 +46,7 @@ struct IdxJoinTest : public TerrierTest {
 
     auto plan = tcop_->OptimizeBoundQuery(common::ManagedPointer(&context_), stmt.ParseResult());
-    if (qtype >= network::QueryType::QUERY_CREATE_TABLE) {
+    if (qtype >= network::QueryType::QUERY_CREATE_TABLE && qtype != network::QueryType::QUERY_CREATE_INDEX) {
```
Nit: It might be better to have a descriptive function / this seems a bit hacky. See NetworkUtil::CreateQueryType.

wuwenw replied:
per our discussion, we agreed to leave it for now

wuwenw commented:
A few major changes since last review:
1. index is created before generating TPL code. 
2. a tablePR  is no longer allocated to set up the indexPR, and values are got from VPI directly.
3. According to 2, functions for init tablePR and fillTablePR are removed from builtins and storageInterface

wuwenw commented on src/execution/compiler/operator/index_create_translator.cpp:
```diff
@@ -0,0 +1,189 @@
+      function->Append(assign);
+      IndexInsert(ctx, function);
+      // Push to parent.
+      // ctx->Push(function);
```
Added comment
